### ISSUE TITLE
Provide matchit-like behavior on % for Ruby delimiters

### ISSUE THREAD

**Issue by CKolkey** [nvim-treesitter/nvim-treesitter#584]
**Describe the bug**
Inline `if`/`unless` is being treated as requiring an `end`, when that is not the case.

**To Reproduce**
With an empty init.nvim, except for treesitter config and `Plug 'andymass/vim-matchup'` to visualize the issue.

Open ruby file:
![Screen Shot 2020-10-14 at 16 34 57](https://user-images.githubusercontent.com/7228095/96004289-801a5200-0e3b-11eb-899e-f817e7a10c50.png)
![Screen Shot 2020-10-14 at 16 34 48](https://user-images.githubusercontent.com/7228095/96004291-814b7f00-0e3b-11eb-8994-96551289e54f.png)

The closing `end` tag for `class` is paired with `def`, and the end tag for `def` is paired with `if`. 

This also happens with inline `unless` statements.

**Expected behavior**
Using `%` to jump between opening/closing do blocks should work. class/end and def/end, ignoring the inline if.

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `git` executable found.
  - OK: `cc` executable found.

## ruby parser healthcheck
  - OK: ruby parser found.
  - OK: `highlights.scm` found.
  - OK: `locals.scm` found.
  - WARNING: No `textobjects.scm` query found for ruby
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter
  - WARNING: No `folds.scm` query found for ruby
    - ADVICE:
      - Open an issue at https://github.com/nvim-treesitter/nvim-treesitter

</code>
</details>

**Output of `nvim --version`**
```
NVIM v0.5.0-752-g759a05407
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-5 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/travis/build/neovim/bot-ci/build/neovim/build/config -I/home/travis/build/neovim/bot-ci/build/neovim/src -I/home/travis/build/neovim/bot-ci/build/neovim/.deps/usr/include -I/usr/include -I/home/travis/build/neovim/bot-ci/build/neovim/build/src/nvim/auto -I/home/travis/build/neovim/bot-ci/build/neovim/build/include
Compiled by travis@travis-job-8293fb89-88d1-426c-a5bd-f2bf6be200ec

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/travis/build/neovim/bot-ci/build/neovim/build/nvim.AppDir/usr/share/nvim
"

```

**Comment by theHamsta** [comment]
Our plugin does not change the behavior of `%` at all. This could be implemented as an additional module for nvim-tree-sitter.

The example code is parsed correctly by tree-sitter-ruby.

**Comment by kyazdani42** [comment]
I've been thinking about that too, would be great because % sometimes doesn't work as expected with vim.
I dont think syntax has anything to do with this, if it does, it might be an upstream issue.

**Comment by theHamsta** [comment]
We could have pairs.scm with start/end for parens and such start end blocks.

**Comment by kyazdani42** [comment]
and that would help for indent too

**Comment by CKolkey** [comment]
Thanks for the reply. I'm still a little confused as to why this _isn't_ treesitter. Without treesitter enabled the matchit/% behavior is as expected, but with it enabled the behavior is incorrect. This holds true with both my full config and a very stripped down config. If what you say is right (and I have no reason to suspect it isn't) and treesitter doesn't change the behavior of %....why does the behavior change with treesitter enabled?

**Comment by theHamsta** [comment]
Does the behavior change in the same way with `set syntax=off`?

**Comment by CKolkey** [comment]
`set syntax=off` changes nothing, unfortunately. With the following init.nvim, I get incorrect matchit/% behavior.

```
call plug#begin()
    Plug 'andymass/vim-matchup'
    Plug 'nvim-treesitter/nvim-treesitter'
call plug#end()

set syntax=off

lua <<EOF
require'nvim-treesitter.configs'.setup {
    ensure_installed = "all",
    highlight = {
        enable = true, 
        disable = { "c", "rust" },
    },
}
EOF
```

**Comment by kyazdani42** [comment]
i believe there might be some vim behavior that is loaded through the syntax option, it behaves the same with lua (% doesn't jump at the right ends). 
That could also be this % works with vim regexes defined in syntax files ?

**Comment by theHamsta** [comment]
The Lua based highlighting and the vim regex highlighting are two different things. Right now there's no real prioritization between the two. The highlight module therefore deactivates `syntax`

**Comment by rafikdraoui** [comment]
> there might be some vim behavior that is loaded through the syntax option

Another visible side-effect of not loading the syntax file is the loss of syntax-based folds (`:help fold-syntax`). But they can be replaced by nvim-treesitter's folds.

**Comment by andymass** [comment]
I'm a bit late to this issue but yes, match-up relies on syntax group information for ruby because the syntax is ambiguous from a myopic standpoint (`end` can match with multiple things, and we need to skip the `if` which is not a block here).  I suspect mine is not the only plugin that breaks when syntax is turned off or unavailable.

I've recently added support for python using tree-sitter https://github.com/andymass/vim-matchup/tree/master/lua.  You can enable it as follows (but will only be active for python right now):

```lua
require'nvim-treesitter.configs'.setup {
  matchup = {
     enable = true
  },
}
```

i.e., it's a real nvim-treesitter plugin.

I can think of a few ways to approach the issue at hand:

1. Use real tree-sitter queries for ruby, like I do with python.  This is the long term plan for most languages, to source accurate document structure information from treesitter.  Unfortunately, it's not trivial to write the queries required to do this, and also it doesn't work for malformed code or code within comments (match-up is capable of correctly matching fully commented code).

2. Allow querying neovims's tree-sitter highlighting result for the syntax groups being applied to the document.  This is like `synID()` but when using the tree-sitter provider.  Is there a plan to "harmonize" the old and new syntax engines somehow?

*TLDR*; Can 2. be done today?  How can I find out what highlighting was applied?

**Comment by vigoux** [comment]
There's plenty of matchit-like plugins now, so I am going to close this issue.
